File: src/lib/api/git/types/conflicts.ts
// In-progress operations + conflicted files — mirrors
// `src-tauri/src/git/types/conflicts.rs`.

/** The active in-progress operation that can stop on conflicts. "none" when the
 * repo is clean / no operation is underway. "carry" is GitLane's worktree-handoff
 * carry (GL-74) — a stash re-apply left conflicts with no git sequencer state. */
export type OperationKind = "merge" | "rebase" | "cherry-pick" | "revert" | "carry" | "none";

/** Non-drivable in-progress git state surfaced as a read-only banner (GitLane
 * can't continue/abort these in-app): `git am` or bisect. "" when the repo is
 * clean or in a drivable operation. */
export type OperationAdvisory = "apply-mailbox" | "bisect" | "";

/** One conflicted (unmerged) path. */
export interface ConflictFile {
  path: string;
  /** "text" (line-mergeable), "binary", or "deleted" (one side removed it). */
  kind: "text" | "binary" | "deleted";
  /** For "deleted", the side that removed the file — "both" when a DD conflict
   * (e.g. rename/rename) left no side with a version; else "". */
  deletedSide: "ours" | "theirs" | "both" | "";
}

/** The in-progress operation + its outstanding conflicts (see Rust
 * `git::conflicts::operation_status`). */
export interface OperationStatus {
  kind: OperationKind;
  /** True when the operation supports skipping the current commit. */
  canSkip: boolean;
  conflicts: ConflictFile[];
  /** A non-drivable in-progress state (git am / bisect) shown as a read-only
   * banner, independent of the drivable `kind`. */
  advisory: OperationAdvisory;
}

/** Raw conflicted content of one text file (with git's merge markers). */
export interface ConflictFileContent {
  path: string;
  content: string;
  binary: boolean;
}

File: src-tauri/src/git/types/conflicts.rs
//! The in-progress merge/sequencer operation and the conflicted files it left.

use serde::Serialize;

/// The in-progress merge/sequencer operation that left the repo in a conflicted
/// or mid-operation state — mapped from libgit2's `RepositoryState`, so a
/// rebase/cherry-pick/revert started from a terminal is detected too. `Carry`
/// is GitLane's own worktree-handoff carry (GL-74), not a git state.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum OperationKind {
    Merge,
    Rebase,
    CherryPick,
    Revert,
    Carry,
    None,
}

/// A non-drivable in-progress git state surfaced as a read-only advisory (not
/// the conflict workspace). These have no in-app continue/abort — the banner
/// points the user at the terminal — so they stay out of [`OperationKind`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum OperationAdvisory {
    ApplyMailbox,
    Bisect,
    /// The repo is clean or in a drivable operation.
    #[serde(rename = "")]
    None,
}

/// Drives the conflict-resolution workflow.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OperationStatus {
    pub kind: OperationKind,
    /// True when the operation supports `--skip` (rebase/cherry-pick/revert).
    pub can_skip: bool,
    /// Unmerged (conflicted) paths still needing resolution. Empty when the
    /// operation has no outstanding conflicts (e.g. all already staged).
    pub conflicts: Vec<ConflictFile>,
    pub advisory: OperationAdvisory,
}

/// How one conflicted (unmerged) path should be resolved in the UI.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ConflictKind {
    /// Both sides changed, line-mergeable.
    Text,
    /// Both sides changed, not line-mergeable (or a symlink/gitlink).
    Binary,
    /// One side (or both) removed the file.
    Deleted,
}

/// One conflicted (unmerged) path in the index.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConflictFile {
    pub path: String,
    pub kind: ConflictKind,
    /// For the [`ConflictKind::Deleted`] kind, which side removed it:
    /// "ours" | "theirs", or "both" for a both-deleted (DD) conflict. Empty
    /// for text/binary conflicts.
    pub deleted_side: String,
}

/// The raw conflicted content of one text file — the worktree copy git wrote
/// with `<<<<<<< / ======= / >>>>>>>` markers — for the in-app editor to parse
/// into hunks. The frontend owns the marker parsing (pure + testable).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConflictFileContent {
    pub path: String,
    pub content: String,
    /// True when the file is binary (no marker content; the editor offers a
    /// whole-file ours/theirs choice instead).
    pub binary: bool,
}
